### PR TITLE
Jpcbertoldo/mvtec ad loco

### LINKED ISSUE #471: Add support for MVTEC LOCO AD
**Is your feature request related to a problem? Please describe.**

MVTec AD dataset only consists of low-level, textural/structural anomalies. To test the capability at detecting high-level, semantic/logical anomalies, additional datasets are needed.

**Describe the solution you'd like**

Integrate a default dataloader and evaluation scheme for [MVTEC LOCO AD](https://link.springer.com/article/10.1007/s11263-022-01578-9#Fn1) dataset

### PR DESCRIPTION
**Pull request by jpcbertoldo**
**Disclaimer**

I am creating this just to keep track of the branch, **please ignore the PR for the moment**.

---

# Description

Create a new dataset: MVTec LOCO Anomaly Detection.

"LOCO" stands for "**LO**gical **CO**nstraints"

I based myself on `anomalib/data/mvtec.py`.

- Fixes #471

## `imread_strategy` 

The dataset supports an option `imread_strategy` which allows the user how to choose when the images are loaded:
  - `onthefly`: behaviour I found in `mvtec.py`, the images are loaded upon demand during the training;
  - `preload`: all the images are cached in the memory (RAM, not GPU) when the dataset is being initialized.
  
## `anotype` and `super_anotype`

Besides providing the binary `label`, I also create the dataset with two other categorical values:
- `super_anotype`: is it a logical or structural anomaly? (or a normal?) 
- `anotype`: "what is the problem with the image?", mvtec ad also has different types of anomalies for each category but this is particularly more interesting here because there are many types of logical violations possible.  

I specifically included this because I am interested in evaluating separately by those types but I will later create an issue for that feature.

## `mask` vs. `masks`

MVTec LOCO's logical anomalies may include several anoamlies in a single image and to properly evaluate them one needs to consider them separately so they are segmented in different mask files in the ground truth.

Since the rest of library expects a tensor `mask` (SINGULAR), I merge them all into a single binary maks (with loss information because they cannot be separated anymore).

In order to later peform proper evaluation there is a second tensor `masks` (PLURAL) which encodes each anomalous region with a different value (0 is a normal pixel, and 1, 2, ..., N are anomalous pixels).

### 1) `self.transform_config_val = self.transform_config_train`

```
        if self.transform_config_train is not None and self.transform_config_val is None:
            self.transform_config_val = self.transform_config_train
```

Is there a good reason for assuming this? 

For me it could make sense that `self.transform_config_val` could have light data augmentations (say, tiny brightness changes) but that should not be repeated in the validation set.

### 2) `split_normal_images_in_train_set(samples, split_ratio, seed)`

MVTec LOCO already defines fixed validation sets so i did not include the option of doing it dinamically like in MVTec AD.

---

# Checklists

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://openvinotoolkit.github.io/anomalib/guides/using_pre_commit.html#pre-commit-hooks) of this project.
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [] New and existing tests pass locally with my changes

### CODE REVIEW

**Comment by jpcbertoldo** [comment]
`MVTec.__init__` has the following lines

```
        if self.transform_config_train is not None and self.transform_config_val is None:
            self.transform_config_val = self.transform_config_train
```

which i did not do in `MVTecLOCO.__init__`.

Is there a good reason for assuming this? For me it could make sense that `self.transform_config_val` could have light data augmentations (say, tiny brightness changes) but that should not be repeated in the validation set.

**Comment by jpcbertoldo** [comment]
another part, in `anomalib.data.mvtec.make_mvtec_dataset`, which i didnt understand why it's there and removed in mvtec loco

```
    # Split the normal images in training set if test set doesn't
    # contain any normal images. This is needed because AUC score
    # cannot be computed based on 1-class
    if sum((samples.split == "test") & (samples.label == "good")) == 0:
        samples = split_normal_images_in_train_set(samples, split_ratio, seed)
```
if that case does happen then there is a violation in the dataset files and an error should be raised

...right?

**Comment by jpcbertoldo** [comment]
yet something i dont understand in `anomalib.data.mvtec`

```
# in MVTecLOCODataset.__getitem__
pre_processed = self.pre_process(image=image, mask=mask)
```

how this works?

does the transform re-apply the last call when mask is not None?

**Comment by jpcbertoldo** [comment]
> Check out this pull request on  [ReviewNB](https://app.reviewnb.com/openvinotoolkit/anomalib/pull/538)
> 
> See visual diffs & provide feedback on Jupyter Notebooks.
> 
> Powered by [ReviewNB](https://www.reviewnb.com/?utm_source=gh)

@samet-akcay is this some sort of integratin in the project?

**Comment by samet-akcay** [comment]
> > Check out this pull request on  [ReviewNB](https://app.reviewnb.com/openvinotoolkit/anomalib/pull/538)
> > See visual diffs & provide feedback on Jupyter Notebooks.
> > Powered by [ReviewNB](https://www.reviewnb.com/?utm_source=gh)
> 
> @samet-akcay is this some sort of integratin in the project?

We use `ReviewNB` for notebook reviews. It's not so easy to review jupyter notebooks on GitHub. ReviewNB makes it significantly easier.

File: anomalib/data/mvtec_loco.py [hunk 1]
```diff
@@ -0,0 +1,863 @@
...
+    # todo correct inconsistency: `transform_config_*val*` used for val and
+    #   test set, but `*test*_batch_size` used for val and set
+
+    def __init__(
+        self,
+        root: str,
+        category: str,
+        task: str = TASK_SEGMENTATION,
+        imread_strategy: str = IMREAD_STRATEGY_PRELOAD,
+        image_size: Optional[Union[int, Tuple[int, int]]] = None,
```
Comment (jpcbertoldo) [inline anomalib/data/mvtec_loco.py:702]: The images in this dataset are not squared.
The ratio of widh/height can end up too different than the original image when the image size is given as an int.

Maybe we should add a warning here?

File: anomalib/data/mvtec_loco.py [hunk 2]
```diff
@@ -0,0 +1,863 @@
...
+
+        else:
+            mask_dict = self._get_masks(index)
+
+        item.update(
+            {
+                "mask_paths": str(self.samples.iloc[index]["mask_paths"]),
+                # TODO CHECK IF THE DOUBLE CALL TO PREPROCESS WILL WORK WITH ALBUMENTATIONS
+                "masks": self.pre_process(image=image, mask=mask_dict["masks"])["mask"],
+                "mask": self.pre_process(image=image, mask=mask_dict["mask"])["mask"],
```
Comment (jpcbertoldo) [inline anomalib/data/mvtec_loco.py:682]: `self.pre_process` is being called for the 3rd time here, will that create any problems?

I'm thinking that maybe the random transforms will apply the same transform every two times (for the image and for the mask).

**Comment by jpcbertoldo** [comment]
i messed some git commands, i'm closing this one and openning another one
